src/main/java/com/amazon/opendistroforelasticsearch/sql/cursor/DefaultCursor.java

```diff
@@ -0,0 +1,217 @@
+
+    /** Make sure all keys are unique to prevent overriding
+     * and as small as possible to make cursor compact
+     */
+    private static final String FETCH_SIZE = "f";
+    private static final String ROWS_LEFT = "l";
+    private static final String INDEX_PATTERN = "i";
+    private static final String SCROLL_ID = "s";
+    private static final String SCHEMA_COLUMNS = "c";
+    private static final String FIELD_ALIAS_MAP = "a";
+
+    /** To get mappings for index to check if type is date needed for
+     * @see com.amazon.opendistroforelasticsearch.sql.executor.format.DateFieldFormatter */
+    private String indexPattern;
+
+    /** List of Schema.Column for maintaining field order and generating null values of missing fields */
+    private List<Schema.Column> columns;
+
+    /** To delegate to correct cursor handler to get next page*/
+    private final CursorType type = CursorType.DEFAULT;
+
+    /**
+     * Truncate the @see DataRows to respect LIMIT clause and/or to identify last page to close scroll context.
+     * docsLeft is decremented by fetch_size for call to get page of result.
+     */
+    private long rowsLeft;
+
+    /** @see com.amazon.opendistroforelasticsearch.sql.executor.format.SelectResultSet */
+    private Map<String, String> fieldAliasMap;
+
+    /** To get next batch of result */
+    private String scrollId;
+
+    /** To reduce the number of rows left by fetchSize */
+    private Integer fetchSize;
+
+
+    private Integer limit;
+
+    @Override
+    public CursorType getType() {
+        return type;
+    }
+
+    @Override
+    public String generateCursorId() {
+        if (rowsLeft <=0 || Strings.isNullOrEmpty(scrollId)) {
+            return null;
+        }
+        JSONObject json = new JSONObject();
+        json.put(FETCH_SIZE, fetchSize);
+        json.put(ROWS_LEFT, rowsLeft);
+        json.put(INDEX_PATTERN, indexPattern);
+        json.put(SCROLL_ID, scrollId);
+        json.put(SCHEMA_COLUMNS, getSchemaAsJson());
+        json.put(FIELD_ALIAS_MAP, fieldAliasMap);
+        return String.format("%s:%s", type.getId(), encodeCursor(json));
+    }
+
+    public static DefaultCursor from(String cursorId) {
+        /**
+         * It is assumed that cursorId here is the second part of the original cursor passed
+         * by the client after removing first part which identifies cursor type
+         */
+         JSONObject json = decodeCursor(cursorId);
+         DefaultCursor cursor = new DefaultCursor();
+         cursor.setFetchSize(json.getInt(FETCH_SIZE));
+         cursor.setRowsLeft(json.getLong(ROWS_LEFT));
+         cursor.setIndexPattern(json.getString(INDEX_PATTERN));
+         cursor.setScrollId(json.getString(SCROLL_ID));
+         cursor.setColumns(getColumnsFromSchema(json.getJSONArray(SCHEMA_COLUMNS)));
+         cursor.setFieldAliasMap(fieldAliasMap(json.getJSONObject(FIELD_ALIAS_MAP)));
+
+         return cursor;
+    }
+
+
+    public String getIndexPattern() {
```

Comment:
minor: consider to use Lombok to auto generate get/set method.

Reply:
Sure, will consider using Lombok.